File: backend/lib/bankfacil/order.rb

module BankFacil
	class Order
		attr_reader :customer, :items, :closed_at
		attr_reader :address
		attr_accessor :voucher

		def initialize(customer:, address:, voucher:)
			@customer = customer
			@address = address
			@voucher = voucher
			@items = []
		end

		def add_product(product, quantity)
			@items << OrderItem.new(self, product, quantity)
		end

		def total_amount
			amount = @items.map(&:total).inject(:+)
			amount = amount - @voucher.amount if @voucher
			amount = 0 if amount < 0 # amount can't be negative
			amount
		end

		def close(closed_at = Time.now)
    		@closed_at = closed_at
    		@voucher.cancel
    		@items.each { |item| item.when_order_closed }
  		end
	end
end

File: backend/BankFacil/Order.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace BankFacil
{
    public class Order
    {
        public Customer Customer { get; }
        public IList<OrderItem> Items { get; }
        public Address Address { get; }
        public DateTime ClosedAt { get; private set; }

        public Order(Customer customer, Address address)
        {
            Customer = customer;
            Items = new List<OrderItem>();
            Address = address ?? new Address("45678-979");
        }

        public void AddProduct(Product product)
        {
            Items.Add(new OrderItem(this, product));
        }

        public int TotalAmount => Items.Sum(x => x.Total);

        public void Close(DateTime? closedAt = null)
        {
            ClosedAt = closedAt ?? DateTime.Now;
        }
    }
}
